The jumping scrubber problem was caused by the playhead position gett…

pcaseooyala opened the pull request:
The jumping scrubber problem was caused by the playhead position getting updated to its old position at the end of the drag. The fix prevents updating the playhead position during the drag and sets the playhead position correctly at the end of the drag.

pcaseooyala committed:
The jumping scrubber problem was caused by the playhead position getting updated to its old position at the end of the drag.  The fix prevents updating the playhead position during the drag and sets the playhead position correctly at the end of the drag.

lental commented on sdk/react/bottomOverlay.js:
```diff
@@ -100,6 +100,12 @@ var BottomOverlay = React.createClass({
 
+  componentWillReceiveProps: function(nextProps) {
+    if (Math.abs(this.props.playhead - nextProps.playhead) >= 0.2) {
+       this.setState({playhead:-1.0}); 
```
can we rename playhead?  it makes the this.state.playhead and this.props.playhead confusing.

maybe something like cachedPlayhead, seeking Playhead, etc.

pcaseooyala replied:
I vote for cachedPlayhead.

lental commented on sdk/react/bottomOverlay.js:
```diff
@@ -100,6 +100,12 @@ var BottomOverlay = React.createClass({
   },
 
+  componentWillReceiveProps: function(nextProps) {
+    if (Math.abs(this.props.playhead - nextProps.playhead) >= 0.2) {
```
can you add a comment here?  is this just an assumption that we send updates that are smaller than .2?  maybe instead you can do this.props.playhead != nextProps.playhead?

pcaseooyala replied:
We can't just do "!=" because then the cachedPlayhead will always be set to -1 and thus never be used.  I propose to define a var scrubPositionDelta = 0.2 and to add a comment before the function.

lental commented:
Do we need to set an InitialState?  Can you look into that, and make sure it is initialized properly (it looks like it should be default -1, right?)?

pcaseooyala committed:
Addressed review comments:
-created a scrubPositionDelta variable for use in componentWillReceiveProps
-added comment to componentWillReceiveProps
-changed "playhead" to "cachedPlayhead"
-initialized cachedPlayhead in getInitialState

pcaseooyala commented:
Pushed the requested changes.  Please re-review.

pcaseooyala committed:
Improved comment

pcaseooyala commented:
We need to have scrubPositionDelta because of timing issues.  Without it, if the scrubber is dragged, the scrubber gets rendered with the new position.  After a clock tick, componentWillReceiveProps gets called, but the props.playhead is still at the old position plus a clock tick.  cachedPlayhead gets set to -1, and the scrubber gets rendered with the props.playhead value which causes the scrubber to jump.  

Using scrubPositionDelta works because in componentWillReceiveProps, cachedPlayhead will only get set to -1 if the difference between the old and new playhead positions is more than just a clock tick (meaning the scrubber got dragged).  If componentWillReceiveProps gets called after a clock tick, cachedPlayhead will not get set to -1 (it will remain at the new, dragged position), and the scrubber will be rendered with the correct position.  Essentially, scrubPositionDelta is being used to detect a drag of the scrubber.

pcaseooyala committed:
ooyalaSkinBridgeListener.js: fixed onSeekStarted to set the playhead to seekend instead of seekstart.
bottomOverlay.js: removed scrubPositionDelta because it is no longer needed.
                             fixed componentWillReceiveProps to check if this.props.playhead equals nextProps.playhead and fixed the function comment

pcaseooyala commented:
Pushed the real fix.  ooyalaSkinBridgeListener.onSeekStarted caused problems by setting the playhead to seekstart instead of seekend.  With that fix, scrubPositionDelta is not needed, and bottomOverlay.componentWillReceiveProps can just compare props.playhead and nextProps.playhead for equality.

Also tested on iOS.

pcaseooyala commented:
Note: On iOS, for live streams, the scrubber jumps during seeks, but we will open a separate ticket for that.

Opened: https://jira.corp.ooyala.com/browse/PBA-4829
